controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -391,30 +394,38 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
     }
 
     @Override
-    CompletableFuture<Integer> createNewTransaction(UUID txId, long timestamp, long leaseExpiryTime, long maxExecutionExpiryTime,
-                                                    long scaleGracePeriod) {
+    CompletableFuture<Pair<Integer, List<Integer>>> createNewTransaction(UUID txId,
+                                                                         long timestamp,
+                                                                         long leaseExpiryTime,
+                                                                         long maxExecutionExpiryTime,
+                                                                         long scaleGracePeriod) {
         Preconditions.checkNotNull(txId);
 
-        final CompletableFuture<Integer> result = new CompletableFuture<>();
+        final CompletableFuture<Pair<Integer, List<Integer>>> result = new CompletableFuture<>();
         final Data<Integer> txnData = new Data<>(
                 new ActiveTxnRecord(timestamp, leaseExpiryTime, maxExecutionExpiryTime, scaleGracePeriod, TxnStatus.OPEN)
                 .toByteArray(), 0);
-        synchronized (txnsLock) {
-            activeTxns.putIfAbsent(txId.toString(), txnData);
-        }
         int epoch = activeEpoch.get();
         synchronized (txnsLock) {
             if (!epochTxnMap.containsKey(epoch)) {
                 result.completeExceptionally(StoreException.create(StoreException.Type.DATA_NOT_FOUND, "epoch"));
             } else {
-                epochTxnMap.compute(epoch, (x, y) -> {
-                    y.add(txId.toString());
-                    return y;
+                getActiveSegments().thenAccept(segments -> {
+                    if (segments.isEmpty()) {
+                        // Stream is sealed
+                        result.completeExceptionally(
+                                StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED, "SEALED"));
+                    } else {
+                        activeTxns.putIfAbsent(txId.toString(), txnData);
+                        epochTxnMap.compute(epoch, (x, y) -> {
```

Comment:
i see a bug in original code.. we should wrap compute in try catch where we complete result in case 'compute' throws an exception.. 
currently if compute throws an exception, then result is never completed.. 

Also, we should have a check for y != null before doing a y.add